eth/peerset.go

```diff
@@ -234,16 +279,74 @@ func (ps *peerSet) waitDiffExtension(peer *eth.Peer) (*diff.Peer, error) {
 	}
 }
 
+// waitTrustExtension blocks until all satellite protocols are connected and tracked
+// by the peerset.
+func (ps *peerSet) waitTrustExtension(peer *eth.Peer) (*trust.Peer, error) {
+	// If the peer does not support a compatible `trust`, don't wait
+	if !peer.RunningCap(trust.ProtocolName, trust.ProtocolVersions) {
+		return nil, nil
+	}
+	// If the peer isn't verify node, don't register trust extension into eth protocol.
+	if !peer.VerifyNode() {
+		return nil, nil
+	}
+	// Ensure nobody can double connect
+	ps.lock.Lock()
+
+	id := peer.ID()
+	if _, ok := ps.peers[id]; ok {
+		ps.lock.Unlock()
+		return nil, errPeerAlreadyRegistered // avoid connections with the same id as existing ones
+	}
+	if _, ok := ps.trustWait[id]; ok {
+		ps.lock.Unlock()
+		return nil, errPeerAlreadyRegistered // avoid connections with the same id as pending ones
+	}
+	// If `trust` already connected, retrieve the peer from the pending set
+	if trust, ok := ps.trustPend[id]; ok {
+		delete(ps.trustPend, id)
+
+		ps.lock.Unlock()
+		return trust, nil
+	}
+	// Otherwise wait for `trust` to connect concurrently
+	wait := make(chan *trust.Peer)
+	ps.trustWait[id] = wait
+	ps.lock.Unlock()
+
+	select {
+	case peer := <-wait:
+		return peer, nil
+
+	case <-time.After(extensionWaitTimeout):
+		ps.lock.Lock()
+		delete(ps.trustWait, id)
+		ps.lock.Unlock()
+		return nil, errPeerWaitTimeout
+	}
+}
+
 func (ps *peerSet) GetDiffPeer(pid string) downloader.IDiffPeer {
 	if p := ps.peer(pid); p != nil && p.diffExt != nil {
 		return p.diffExt
 	}
 	return nil
 }
 
+// GetVerifyPeers returns an array of verify nodes.
+func (ps *peerSet) GetVerifyPeers() []*trustPeer {
+	res := make([]*trustPeer, 0)
```

Comment:
Race condition on `peerSet.peers`, try avoid that.

Reply:
You are right, I will add rlock here.